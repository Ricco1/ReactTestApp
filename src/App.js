import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { getImagesAction } from "./actions/getImages";
import { hideNotification } from "./actions/notifications";
import PropTypes from 'prop-types';
import Pagination from './components/Pagination';
import NotificationBar from './components/NotificationBar';
import logo from './logo.svg';
import './App.css';

const List = ({image}) => (
    <div className='col-3 img-container'>
        <img className='img-fluid' src={image.images.fixed_height_still.url} alt="gif"/>
    </div>
);

export class App extends PureComponent {
    static propTypes = {
        images: PropTypes.array
    };

    static defaultProps = {
        images: [],
        isFetching: false,
        notifications:""
    };

    constructor() {
        super();

        this.state = {
            imagesToShow: [],
            imagesToShowInPage: 8,
            paginationActivePage: 1,
            paginationPagesTotal: 0,
            totalMbs: 0
        };

        this.nextPage = this.nextPage.bind(this);
        this.prevPage = this.prevPage.bind(this);
        this.selectPage = this.selectPage.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.switchPage(nextProps, this.state)
    }

    switchPage(props, state, direction, page) {
        if(props.images.length) {
            const {paginationActivePage, imagesToShowInPage} = state;
            let paginationPagesTotal = Math.ceil(props.images.length / imagesToShowInPage),
                increment,
                targetPage,
                totalMbs;

            increment = direction ? (direction === 'next' ? 1 : -1) : 0;
            targetPage = page ? page : paginationActivePage + increment;

            if(targetPage > paginationPagesTotal || targetPage < 1) return;


            let startingIndex = (targetPage - 1) * imagesToShowInPage,
                imagesToSlice = (startingIndex + imagesToShowInPage) < props.images.length  ?
                                startingIndex + imagesToShowInPage :
                                (startingIndex > props.images.length) ? 0 : props.images.length,

                imagesToShow = props.images.slice(startingIndex, imagesToSlice);

            totalMbs = this.calculatePictureSize(imagesToShow);

            this.setState({
                imagesToShow,
                paginationPagesTotal,
                paginationActivePage: targetPage,
                totalMbs
            });
        }
    }

    calculatePictureSize(imageArray) {
        let size = 0,
            currentImageSize = 0,
            byteInMB = 0.00000095367432;

        for(let j = 0, jlen = imageArray.length; j < jlen; j++) {
            currentImageSize = imageArray[j].images.fixed_height_still.size ? +imageArray[j].images.fixed_height_still.size : 0;
            size += currentImageSize;
        }

        return size * byteInMB;
    }

    nextPage() {
        const {props, state} = this;
        this.switchPage(props, state, 'next')
    }

    prevPage() {
        const {props, state} = this;
        this.switchPage(props, state, 'prev')
    }

    selectPage(e,pageNumber) {
        e && e.preventDefault();
        const {props, state} = this;
        if(pageNumber === state.paginationActivePage) return;
        this.switchPage(props, state, null, pageNumber);
    }

    render() {
        const {imagesToShow, paginationPagesTotal, paginationActivePage, totalMbs} = this.state;
        const {onGetImagesClick, notifications, hideNotification} = this.props;
        return (
            <div className="App">
                {notifications.length ? <NotificationBar notification={notifications} hide={hideNotification} /> : ''}
                <header className="App-header">
                    <img src={logo} className="App-logo" alt="logo" />
                    <h1 className="App-title">Welcome to React!!</h1>
                </header>
                <p className="App-intro mt-2 mb-2">
                    <button onClick={onGetImagesClick}>Get images</button>
                </p>
                { this.props.isFetching ? <p className='loader text-center'>Loading...</p> : '' }
                <div className='container'>
                    {
                        imagesToShow.length > 0 ?
                            <div className='row'>
                                <div className='col d-flex align-items-center'>
                                    <div className='size-count mr-auto'>{totalMbs} MB Total</div>
                                    <Pagination pagesTotal={paginationPagesTotal} currentPage={paginationActivePage} nextPage={this.nextPage} prevPage={this.prevPage} selectPage={this.selectPage} />
                                </div>
                            </div> : ''
                    }
                    <div className='row'>
                        { imagesToShow.map((image, i) => <List key={i} image={image} /> )}
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = ({images, isFetching, notifications}) => ({
    images,
    isFetching,
    notifications
});

const mapDispatchToProps = {
    onGetImagesClick: getImagesAction,
    hideNotification
};

export default connect(mapStateToProps,mapDispatchToProps)(App);
