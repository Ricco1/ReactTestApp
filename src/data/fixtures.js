export const imagesArray  = [
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    },
    {
        "images": {
            "fixed_height_still": {
                "url": "https://media2.giphy.com/media/8lZkBk00V3ecftf8aN/200_s.gif?cid=e1bb72ff5abf5a4e5078307859d2e456",
                "width": "328",
                "height": "200",
                "size": "39564"
            }
        },
        "title": "empty kate mckinnon GIF by Saturday Night Live"
    }
];

export const imagesData = {
    data: imagesArray
};