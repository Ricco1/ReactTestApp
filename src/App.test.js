import React from 'react';
import { shallow, mount } from 'enzyme';
import { imagesArray } from './data/fixtures';
import  { App } from './App';

const props = {
    images: [],
    isFetching: false,
    notifications: ''
};

describe('App', () => {
    let appWrapper = mount(<App {...props} />);

    it('renders the App', () => {
        expect(appWrapper).toMatchSnapshot();
    });

    describe('Image loader', () => {
        beforeEach(() => {
            appWrapper.setProps({images: [], isFetching: true, notifications: ''});
        });

        it('shows `Is Loading...` while images are fetching', () => {
            expect(appWrapper.find('.loader').exists()).toBe(true);
        })
    });

    describe('Notification bar', () => {
        beforeEach(() => {
            appWrapper.setProps({images: [], isFetching: false, notifications: 'Error message'});
        });

        it('shows Notification bar', () => {
            expect(appWrapper.find('NotificationBar').exists()).toBe(true)
        });

        it('shows Notification bar with message', () => {
            const noteBarText = appWrapper.find('NotificationBar .alert').text();
            expect(noteBarText.substring(0,noteBarText.length - 1)).toEqual('Error message');
        })
    });

    describe('and renders images', () => {
        beforeEach(() => {
            appWrapper.setProps({images: imagesArray, isFetching: false, notifications: ''});
        });

        it('with correct number of MB per page', () => {
           // console.log(appWrapper.state())
            const currentPage = appWrapper.state().paginationActivePage,
                maxImagesOnPage = appWrapper.state().imagesToShowInPage,
                startingIndex = (currentPage - 1) * maxImagesOnPage,
                imagesToSlice = maxImagesOnPage * currentPage,
                currentPageImg = imagesArray.slice(startingIndex, imagesToSlice);

            expect(appWrapper.find('.size-count').text()).toEqual(appWrapper.instance().calculatePictureSize(currentPageImg) + ' MB Total');
        });

        it('with correct number of pages', () => {
          //console.log(appWrapper.debug())
            expect(appWrapper.find('Pagination li a').length).toEqual(appWrapper.state().paginationPagesTotal);
        });

        it('with correct number of images', () => {
          expect(appWrapper.find('.img-container img').length).toEqual(appWrapper.state().imagesToShow.length)
        });
    });

    describe('when clicking the `Next` Pagination button', () => {
        beforeEach(() => {
            const nextButton = appWrapper.find('span.page-link').at(1);
            nextButton.simulate('click');
        });
        afterEach(() => {
            const prevButton = appWrapper.find('span.page-link').at(0);
            prevButton.simulate('click');
        });
        it('will render next Pagination page', () => {
            expect(appWrapper.state().paginationActivePage).toEqual(2);
            expect(appWrapper.find('.img-container img').length).toEqual(appWrapper.state().imagesToShow.length)
        });

        it('will have `active` class on pagination button with number 2', () => {
            expect(appWrapper.find('.page-item.active a').text()).toEqual('2');
        });
    });

    describe('when clicking non-active Pagination button', () => {
        beforeEach(() => {
            const secondButton = appWrapper.find('.pagination li.page-item').at(3);
            secondButton.simulate('click');
        });
        afterEach(() => {
            const firstButton = appWrapper.find('.pagination li.page-item').at(2);
            firstButton.simulate('click');
        });
        it('will select the appropriate page', () => {
            expect(appWrapper.find('.page-item.active a').text()).toEqual('2');
        });
    })
});