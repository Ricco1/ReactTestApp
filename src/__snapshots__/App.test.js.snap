// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the App 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Welcome to React!!
    </h1>
  </header>
  <p
    className="App-intro mt-2 mb-2"
  >
    <button>
      Get images
    </button>
  </p>
  <div
    className="container"
  >
    <div
      className="row"
    />
  </div>
</div>
`;
